### PR TITLE
PCP: add pd-backup tool

### PR DESCRIPTION
**Pull request by mantuliu**
PCP #1793 
### What problem does this PR solve? 
backup like this:
- clusterID
- max allocated ID
- max allocated timestamp
- configurations

### What is changed and how it works?
add backup tool.

### Check List 

Tests 

 - Manual test

### CODE REVIEW

File: tools/pd-backup/main.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
+package main
```
Comment (disksing) [inline tools/pd-backup/main.go]: Add license header.

File: tools/pd-backup/main.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"github.com/pingcap/pd/pkg/etcdutil"
```
Comment (disksing) [inline tools/pd-backup/main.go]: put standard pkgs and others in 2 groups.

File: tools/pd-backup/main.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+}
+
+func outputToFile(backOut *backupInfo, f *os.File) {
+	w := bufio.NewWriter(f)
+	defer w.Flush()
+	clusterIDOut := "clusterID:" + strconv.FormatUint(backOut.clusterID, 10)
```
Comment (disksing) [inline tools/pd-backup/main.go]: Maybe use JSON or TOML file so later `pd-recover` can read it easier?

**Comment by disksing** [comment]
Could you attach an output example?

File: tools/pd-backup/main.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+	rootPath := path.Join(pdRootPath, strconv.FormatUint(clusterID, 10))
+	allocIDPath := path.Join(rootPath, "alloc_id")
+	resp, err = etcdutil.EtcdKVGet(client, allocIDPath)
+	checkErr(err)
+	allocIDMax, err := typeutil.BytesToUint64(resp.Kvs[0].Value)
+	checkErr(err)
+	backOut.allocIDMax = allocIDMax
+
+	timestapPath := path.Join(rootPath, "timestamp")
```
Comment (rleungx) [inline tools/pd-backup/main.go]:
```suggestion
	timestampPath := path.Join(rootPath, "timestamp")
```

File: tools/pd-backup/main.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	})
+	checkErr(err)
+
+	backOut := getBackupInfo(client)
+	outputToFile(backOut, f)
+	fmt.Println("pd backup successful! dump file is:", *filePath)
+}
+
+func getBackupInfo(client *clientv3.Client) *backupInfo {
+	backOut := &backupInfo{}
```
Comment (lhy1024) [inline tools/pd-backup/main.go]: `backOut` is ambiguous. how about `info` or `backupInfo`

File: tools/pd-backup/main.go (hunk 1 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: done

File: tools/pd-backup/main.go (hunk 4 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: done

File: tools/pd-backup/main.go (hunk 5 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: OK,let me try to add test.

File: tools/pd-backup/main.go (hunk 4 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: done

> Could you attach an output example?
```
{
    "clusterID": 6772158714662687507,
    "allocIDMax": 5000,
    "allocTimestampMax": 1577333080791401000,
    "config": {
        "client-urls": "http://127.0.0.1:2379",
        "peer-urls": "http://127.0.0.1:2380",
        "advertise-client-urls": "http://127.0.0.1:2379",
        "advertise-peer-urls": "http://127.0.0.1:2380",
        "name": "pd-liuguangliangdeMacBook-Pro.local",
        "data-dir": "default.pd-liuguangliangdeMacBook-Pro.local",
        "force-new-cluster": false,
        "enable-grpc-gateway": true,
        "initial-cluster": "pd-liuguangliangdeMacBook-Pro.local=http://127.0.0.1:2380",
        "initial-cluster-state": "new",
        "join": "",
        "lease": 3,
        "log": {
            "level": "",
            "format": "text",
            "disable-timestamp": false,
            "file": {
                "filename": "",
                "max-size": 0,
                "max-days": 0,
                "max-backups": 0
            },
            "development": false,
            "disable-caller": false,
            "disable-stacktrace": false,
            "disable-error-verbose": true,
            "sampling": null
        },
        "log-file": "",
        "log-level": "",
        "tso-save-interval": "3s",
        "metric": {
            "job": "pd-liuguangliangdeMacBook-Pro.local",
            "address": "",
            "interval": "15s"
        },
        "schedule": {
            "max-snapshot-count": 3,
            "max-pending-peer-count": 16,
            "max-merge-region-size": 20,
            "max-merge-region-keys": 200000,
            "split-merge-interval": "1h0m0s",
            "enable-one-way-merge": "false",
            "enable-cross-table-merge": "false",
            "patrol-region-interval": "100ms",
            "max-store-down-time": "30m0s",
            "leader-schedule-limit": 4,
            "leader-schedule-strategy": "\"count\"",
            "region-schedule-limit": 2048,
            "replica-schedule-limit": 64,
            "merge-schedule-limit": 8,
            "hot-region-schedule-limit": 4,
            "hot-region-cache-hits-threshold": 3,
            "store-balance-rate": 15,
            "tolerant-size-ratio": 0,
            "low-space-ratio": 0.8,
            "high-space-ratio": 0.6,
            "scheduler-max-waiting-operator": 3,
            "enable-remove-down-replica": "true",
            "enable-replace-offline-replica": "true",
            "enable-make-up-replica": "true",
            "enable-remove-extra-replica": "true",
            "enable-location-replacement": "true",
            "enable-debug-metrics": "false",
            "schedulers-v2": [
                {
                    "type": "balance-region",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "balance-leader",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "hot-region",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "label",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                }
            ],
            "schedulers": {
                "balance-hot-region-scheduler": "null",
                "balance-leader-scheduler": "{\"name\":\"balance-leader-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}",
                "balance-region-scheduler": "{\"name\":\"balance-region-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}",
                "label-scheduler": "{\"name\":\"label-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}"
            }
        },
        "replication": {
            "max-replicas": 3,
            "location-labels": "",
            "strictly-match-label": "false",
            "EnablePlacementRules": false
        },
        "pd-server": {
            "use-region-storage": "true",
            "max-reset-ts-gap": 86400000000000,
            "key-type": "table",
            "runtime-services": "",
            "metric-storage": ""
        },
        "cluster-version": "1.0.0",
        "quota-backend-bytes": "0B",
        "auto-compaction-mode": "periodic",
        "auto-compaction-retention-v2": "1h",
        "TickInterval": "500ms",
        "ElectionInterval": "3s",
        "PreVote": true,
        "security": {
            "cacert-path": "",
            "cert-path": "",
            "key-path": ""
        },
        "label-property": {},
        "WarningMsgs": null,
        "DisableStrictReconfigCheck": false,
        "HeartbeatStreamBindInterval": "1m0s",
        "LeaderPriorityCheckInterval": "1m0s"
    }
}

```

**Comment by mantuliu** [comment]
> Could you attach an output example?

```
{
    "clusterID": 6772158714662687507,
    "allocIDMax": 5000,
    "allocTimestampMax": 1577333080791401000,
    "config": {
        "client-urls": "http://127.0.0.1:2379",
        "peer-urls": "http://127.0.0.1:2380",
        "advertise-client-urls": "http://127.0.0.1:2379",
        "advertise-peer-urls": "http://127.0.0.1:2380",
        "name": "pd-liuguangliangdeMacBook-Pro.local",
        "data-dir": "default.pd-liuguangliangdeMacBook-Pro.local",
        "force-new-cluster": false,
        "enable-grpc-gateway": true,
        "initial-cluster": "pd-liuguangliangdeMacBook-Pro.local=http://127.0.0.1:2380",
        "initial-cluster-state": "new",
        "join": "",
        "lease": 3,
        "log": {
            "level": "",
            "format": "text",
            "disable-timestamp": false,
            "file": {
                "filename": "",
                "max-size": 0,
                "max-days": 0,
                "max-backups": 0
            },
            "development": false,
            "disable-caller": false,
            "disable-stacktrace": false,
            "disable-error-verbose": true,
            "sampling": null
        },
        "log-file": "",
        "log-level": "",
        "tso-save-interval": "3s",
        "metric": {
            "job": "pd-liuguangliangdeMacBook-Pro.local",
            "address": "",
            "interval": "15s"
        },
        "schedule": {
            "max-snapshot-count": 3,
            "max-pending-peer-count": 16,
            "max-merge-region-size": 20,
            "max-merge-region-keys": 200000,
            "split-merge-interval": "1h0m0s",
            "enable-one-way-merge": "false",
            "enable-cross-table-merge": "false",
            "patrol-region-interval": "100ms",
            "max-store-down-time": "30m0s",
            "leader-schedule-limit": 4,
            "leader-schedule-strategy": "\"count\"",
            "region-schedule-limit": 2048,
            "replica-schedule-limit": 64,
            "merge-schedule-limit": 8,
            "hot-region-schedule-limit": 4,
            "hot-region-cache-hits-threshold": 3,
            "store-balance-rate": 15,
            "tolerant-size-ratio": 0,
            "low-space-ratio": 0.8,
            "high-space-ratio": 0.6,
            "scheduler-max-waiting-operator": 3,
            "enable-remove-down-replica": "true",
            "enable-replace-offline-replica": "true",
            "enable-make-up-replica": "true",
            "enable-remove-extra-replica": "true",
            "enable-location-replacement": "true",
            "enable-debug-metrics": "false",
            "schedulers-v2": [
                {
                    "type": "balance-region",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "balance-leader",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "hot-region",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                },
                {
                    "type": "label",
                    "args": null,
                    "disable": false,
                    "args-payload": ""
                }
            ],
            "schedulers": {
                "balance-hot-region-scheduler": "null",
                "balance-leader-scheduler": "{\"name\":\"balance-leader-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}",
                "balance-region-scheduler": "{\"name\":\"balance-region-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}",
                "label-scheduler": "{\"name\":\"label-scheduler\",\"ranges\":[{\"start-key\":\"\",\"end-key\":\"\"}]}"
            }
        },
        "replication": {
            "max-replicas": 3,
            "location-labels": "",
            "strictly-match-label": "false",
            "EnablePlacementRules": false
        },
        "pd-server": {
            "use-region-storage": "true",
            "max-reset-ts-gap": 86400000000000,
            "key-type": "table",
            "runtime-services": "",
            "metric-storage": ""
        },
        "cluster-version": "1.0.0",
        "quota-backend-bytes": "0B",
        "auto-compaction-mode": "periodic",
        "auto-compaction-retention-v2": "1h",
        "TickInterval": "500ms",
        "ElectionInterval": "3s",
        "PreVote": true,
        "security": {
            "cacert-path": "",
            "cert-path": "",
            "key-path": ""
        },
        "label-property": {},
        "WarningMsgs": null,
        "DisableStrictReconfigCheck": false,
        "HeartbeatStreamBindInterval": "1m0s",
        "LeaderPriorityCheckInterval": "1m0s"
    }
}

```

File: tools/pd-backup/main.go (hunk 3 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: JSON output.PTAL.thanks. @disksing

File: tools/pd-backup/main.go (hunk 4 above)
Comment (mantuliu) [reply tools/pd-backup/main.go]: done

File: tests/pdbackup/backup_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package pdbackup
```
Comment (lhy1024) [inline tests/pdbackup/backup_test.go]: add license please.

File: tests/pdbackup/backup_test.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+var _ = Suite(&backupTestSuite{})
+
+type backupTestSuite struct{}
+
+func (s *backupTestSuite) SetUpSuite(c *C) {
+	server.EnableZap = true
+}
+
+func (s *backupTestSuite) TestBackup(c *C) {
```
Comment (lhy1024) [inline tests/pdbackup/backup_test.go]: Good. do we need to add test to restore from backup info?

Comment (mantuliu) [reply tests/pdbackup/backup_test.go]: @lhy1024 PTAL,thanks.

**Pull request merged**